docs/DeveloperGuide.md

```diff
@@ -0,0 +1,123 @@
+
+### Setting up the project in your computer using IntelliJ
+1. Fork this repo, and clone the fork to your computer.
+2. Open *IntelliJ* (if you are not in the welcome screen, click `File` > `Close Project` to close the existing project dialog first).
+3. Set up the correct *JDK* version for *Gradle*.
+4. Click `Configure` > `Project Defaults` > `Project Structure`.
+5. Click `New…​` and find the directory of the *JDK*.
+6. Click `Import Project`.
+7. Locate the `build.gradle` file and select it. Click `OK`.
+8. Ensure that the selected version of `Gradle JVM` matches our prerequisite.
+9. Click `OK` to accept the all the other default settings.
+
+### Verifying the setup
+1. Ensure that *Gradle* build without error.
+2. Run the tests to ensure they all pass.
+   1. On the project root directory, run the command `gradlew clean build` first, followed by `gradlew functional` and ensure that both builds are successful.
+
+### Configuring the coding style
+This project follows [oss-generic coding standards](https://oss-generic.github.io/process/docs/CodingStandards.html). *IntelliJ’s* default style is mostly compliant with our *Java* coding convention but it uses a different import order from ours. To rectify,
+
+1. Go to `File` > `Settings…`​ (*Windows/Linux*), or `IntelliJ IDEA` > `Preferences…`​ (*macOS*).
+2. Select `Editor` > `Code Style` > `Java`.
+3. Click on the `Imports` tab to set the order
+   * For `Class count to use import with '*'` and `Names count to use static import with '*'`: Set to `999` to prevent IntelliJ from contracting the import statements
+   * For `Import Layout`, follow this image below:
+   ![import-order](images/import-order.png)
+
+Optionally, you can follow the [Using Checkstyle](UsingCheckstyle.md) document to configure *Intellij* to check style-compliance as you write code.
+
+### Before writing code
+1. Do check out our [process guide](../docs/Process.md) before submitting any PR with your changes.
+
+## Architecture
+
+ ![architecture](images/architecture.png)
+*Figure 1. Overall architecture of RepoSense*
+
+### Parser(ConfigParser)
+`Parser` contains two classes:
+ * [`ArgsParser`](/src/main/java/reposense/parser/ArgsParser.java): Parses the user-supplied command line arguments into a `CliArguments` object.
+ * [`CsvParser`](/src/main/java/reposense/parser/CsvParser.java): Parses the the user-supplied CSV config file into a list of `RepoConfiguration` for each repository to analyze.
+
+
+### Git
+`Git` contains the wrapper classes for respective *git* commands.
+ * [`GitDownloader`](/src/main/java/reposense/git/GitDownloader.java): Wrapper class for `git clone` functionality. Clones the repository from *GitHub* into a temporary folder in order to run the analysis.
+ * [`GitChecker`](/src/main/java/reposense/git/GitChecker.java): Wrapper class for `git checkout` functionality. Checks out the repository by branch name or commit hash.
```

Comment:
Note for future: the name `...Checker` may be misleading as it's not the same as 'checkout'.